File: src/services/molecules/AgentConnectionManager.ts
/**
 * AgentConnectionManager - Molecular Component
 * Combines agent service atoms into a functional connection management system
 * Groups of atoms bonded together that take on distinct new properties
 */

import { agentService, AgentConnection, CreateConnectionRequest, AgentDeployment } from '../atoms/AgentService';
import { authFlowManager, AuthFlowConfig } from './AuthFlowManager';

export interface AgentConnectionConfig {
  agentId: string;
  provider: string;
  scopes: string[];
  userId: string;
  providerAccountLabel?: string;
  isDefault?: boolean;
}

export interface AgentConnectionResult {
  success: boolean;
  connection?: AgentConnection;
  deployment?: AgentDeployment;
  error?: string;
}

export interface AgentManagementConfig {
  name: string;
  description?: string;
  agentId: string;
  connections: AgentConnectionConfig[];
  deploymentConfig?: Record<string, any>;
}

/**
 * Molecular agent connection manager
 * Combines atomic agent operations with authentication flows
 * Follows single responsibility principle for connection management
 */
export class AgentConnectionManager {
  private readonly agentService: typeof agentService;
  private readonly authFlowManager: typeof authFlowManager;

  constructor(
    agentServiceInstance = agentService,
    authFlowManagerInstance = authFlowManager
  ) {
    this.agentService = agentServiceInstance;
    this.authFlowManager = authFlowManagerInstance;
  }

  /**
   * Create complete agent connection with authentication
   * Molecular operation: combines auth flow with connection creation
   */
  async createAgentConnection(config: AgentConnectionConfig): Promise<AgentConnectionResult> {
    try {
      // Step 1: Authenticate with the provider
      const authConfig: AuthFlowConfig = {
        provider: config.provider,
        scopes: config.scopes,
        userId: config.userId,
        agentId: config.agentId,
        isDefault: config.isDefault,
      };

      const authResult = await this.authFlowManager.executeAuthFlow(authConfig);

      if (!authResult.success || authResult.status !== 'completed') {
        return {
          success: false,
          error: authResult.error || `Authentication failed with status: ${authResult.status}`,
        };
      }

      if (!authResult.token) {
        return {
          success: false,
          error: 'Authentication succeeded but no token received',
        };
      }

      // Step 2: Create the agent connection
      const connectionRequest: CreateConnectionRequest = {
        oauthTokenId: authResult.token,
      };

      const connection = await this.agentService.createConnection(config.agentId, connectionRequest);

      return {
        success: true,
        connection,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to create agent connection',
      };
    }
  }

  /**
   * Setup complete agent with multiple connections
   * Molecular operation: combines multiple connection creations
   */
  async setupAgent(config: AgentManagementConfig): Promise<AgentConnectionResult> {
    try {
      const connections: AgentConnection[] = [];
      let lastError: string | undefined;

      // Create all connections
      for (const connectionConfig of config.connections) {
        const result = await this.createAgentConnection(connectionConfig);
        
        if (result.success && result.connection) {
          connections.push(result.connection);
        } else {
          lastError = result.error;
          // Continue trying other connections even if one fails
        }
      }

      if (connections.length === 0) {
        return {
          success: false,
          error: lastError || 'Failed to create any connections',
        };
      }

      // Create deployment if config provided
      let deployment: AgentDeployment | undefined;
      if (config.deploymentConfig) {
        try {
          deployment = await this.agentService.createDeployment({
            name: config.name,
            description: config.description,
            ...config.deploymentConfig,
          });
        } catch (error) {
          // Don't fail the whole operation if deployment fails
          console.warn('Deployment creation failed:', error);
        }
      }

      return {
        success: true,
        connection: connections[0], // Return first connection as primary
        deployment,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to setup agent',
      };
    }
  }

  /**
   * Get all connections for an agent
   * Molecular operation: combines connection listing with filtering
   */
  async getAgentConnections(agentId: string): Promise<AgentConnection[]> {
    try {
      return await this.agentService.listConnections(agentId);
    } catch (error) {
      throw new Error(`Failed to get agent connections: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Remove agent connection and clean up authentication
   * Molecular operation: combines connection removal with auth cleanup
   */
  async removeAgentConnection(agentId: string, connectionId: string, userId: string): Promise<boolean> {
    try {
      // Get connection details before removal for cleanup
      const connections = await this.agentService.listConnections(agentId);
      const connection = connections.find(c => c.id === connectionId);
      
      // Remove the connection
      const removed = await this.agentService.removeConnection(agentId, connectionId);
      
      // Clean up authentication if connection was found
      if (connection && removed) {
        await this.authFlowManager.revokeAuthentication(userId, connection.providerId);
      }
      
      return removed;
    } catch (error) {
      return false;
    }
  }

  /**
   * Refresh agent connection
   * Molecular operation: combines connection refresh with auth refresh
   */
  async refreshAgentConnection(
    agentId: string, 
    connectionId: string, 
    userId: string
  ): Promise<AgentConnectionResult> {
    try {
      // Get current connection details
      const connections = await this.agentService.listConnections(agentId);
      const connection = connections.find(c => c.id === connectionId);
      
      if (!connection) {
        return {
          success: false,
          error: 'Connection not found',
        };
      }

      // Refresh authentication
      const authConfig: AuthFlowConfig = {
        provider: connection.providerId,
        scopes: connection.scopes,
        userId,
        agentId,
        isDefault: true,
      };

      const authResult = await this.authFlowManager.refreshToken(authConfig);

      if (!authResult.success || authResult.status !== 'completed') {
        return {
          success: false,
          error: authResult.error || `Authentication refresh failed with status: ${authResult.status}`,
        };
      }

      // Remove old connection and create new one
      await this.agentService.removeConnection(agentId, connectionId);
      
      const connectionRequest: CreateConnectionRequest = {
        oauthTokenId: authResult.token!,
      };

      const newConnection = await this.agentService.createConnection(agentId, connectionRequest);

      return {
        success: true,
        connection: newConnection,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to refresh agent connection',
      };
    }
  }

  /**
   * Get agent deployment status
   * Molecular operation: combines deployment retrieval with status checking
   */
  async getAgentDeploymentStatus(deploymentId: string): Promise<AgentDeployment | null> {
    try {
      return await this.agentService.getDeployment(deploymentId);
    } catch (error) {
      return null;
    }
  }

  /**
   * Delete agent and all its connections
   * Molecular operation: combines multiple deletions with cleanup
   */
  async deleteAgent(agentId: string, userId: string): Promise<boolean> {
    try {
      // Get all connections
      const connections = await this.agentService.listConnections(agentId);
      
      // Remove all connections and clean up auth
      let allConnectionsRemoved = true;
      for (const connection of connections) {
        const removed = await this.removeAgentConnection(agentId, connection.id, userId);
        if (!removed) {
          allConnectionsRemoved = false;
        }
      }

      return allConnectionsRemoved;
    } catch (error) {
      return false;
    }
  }
}

// Export singleton instance
export const agentConnectionManager = new AgentConnectionManager();

File: src/services/molecules/agent_connection_manager.py
"""
Agent Connection Manager - Molecular Component
Combines agent service atoms into a functional connection management system
Groups of atoms bonded together that take on distinct new properties
"""

from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from pydantic import BaseModel, Field

from ..atoms.agent_service import (
    AgentService,
    AgentConnection,
    CreateConnectionRequest,
    AgentDeployment,
    AgentServiceError
)
from ..atoms.auth_service import OAuthToken
from .auth_flow_manager import (
    AuthFlowManager,
    AuthFlowConfig,
    AuthFlowResult,
    AuthFlowManagerError
)


class AgentConnectionConfig(BaseModel):
    """Agent connection configuration"""
    agent_id: str
    provider: str
    scopes: List[str]
    user_id: str
    provider_account_label: Optional[str] = None
    is_default: bool = False


class AgentConnectionResult(BaseModel):
    """Agent connection result"""
    success: bool
    connection: Optional[AgentConnection] = None
    deployment: Optional[AgentDeployment] = None
    error: Optional[str] = None


class AgentManagementConfig(BaseModel):
    """Complete agent management configuration"""
    name: str
    description: Optional[str] = None
    agent_id: str
    connections: List[AgentConnectionConfig] = Field(default_factory=list)
    deployment_config: Dict[str, Any] = Field(default_factory=dict)


class AgentConnectionManagerError(Exception):
    """Custom exception for agent connection manager errors"""
    pass


class AgentConnectionManager:
    """
    Molecular agent connection manager
    Combines atomic agent operations with authentication flows
    Follows single responsibility principle for connection management
    """
    
    def __init__(
        self,
        agent_service: Optional[AgentService] = None,
        auth_flow_manager: Optional[AuthFlowManager] = None
    ):
        self.agent_service = agent_service or AgentService()
        self.auth_flow_manager = auth_flow_manager or AuthFlowManager()
    
    async def create_agent_connection(self, config: AgentConnectionConfig) -> AgentConnectionResult:
        """
        Create complete agent connection with authentication
        Molecular operation: combines auth flow with connection creation
        """
        try:
            # Step 1: Authenticate with the provider
            auth_config = AuthFlowConfig(
                provider=config.provider,
                scopes=config.scopes,
                user_id=config.user_id,
                agent_id=config.agent_id,
                is_default=config.is_default,
            )
            
            auth_result = await self.auth_flow_manager.execute_auth_flow(auth_config)
            
            if not auth_result.success or auth_result.status != "completed":
                return AgentConnectionResult(
                    success=False,
                    error=auth_result.error or f"Authentication failed with status: {auth_result.status}",
                )
            
            if not auth_result.token:
                return AgentConnectionResult(
                    success=False,
                    error="Authentication succeeded but no token received",
                )
            
            # Step 2: Create the agent connection
            connection_request = CreateConnectionRequest(
                oauth_token_id=auth_result.token
            )
            
            connection = await self.agent_service.create_connection(config.agent_id, connection_request)
            
            return AgentConnectionResult(
                success=True,
                connection=connection,
            )
            
        except (AuthFlowManagerError, AgentServiceError) as e:
            return AgentConnectionResult(
                success=False,
                error=str(e),
            )
        except Exception as e:
            return AgentConnectionResult(
                success=False,
                error=f"Failed to create agent connection: {str(e)}",
            )
    
    async def setup_agent(self, config: AgentManagementConfig) -> AgentConnectionResult:
        """
        Setup complete agent with multiple connections
        Molecular operation: combines multiple connection creations
        """
        try:
            connections: List[AgentConnection] = []
            last_error: Optional[str] = None
            
            # Create all connections
            for connection_config in config.connections:
                result = await self.create_agent_connection(connection_config)
                
                if result.success and result.connection:
                    connections.append(result.connection)
                else:
                    last_error = result.error
                    # Continue trying other connections even if one fails
            
            if not connections:
                return AgentConnectionResult(
                    success=False,
                    error=last_error or "Failed to create any connections",
                )
            
            # Create deployment if config provided
            deployment: Optional[AgentDeployment] = None
            if config.deployment_config:
                try:
                    deployment_config = {
                        "name": config.name,
                        "description": config.description,
                        **config.deployment_config,
                    }
                    deployment = await self.agent_service.create_deployment(deployment_config)
                except Exception as e:
                    # Don't fail the whole operation if deployment fails
                    print(f"Warning: Deployment creation failed: {str(e)}")
            
            return AgentConnectionResult(
                success=True,
                connection=connections[0],  # Return first connection as primary
                deployment=deployment,
            )
            
        except Exception as e:
            return AgentConnectionResult(
                success=False,
                error=f"Failed to setup agent: {str(e)}",
            )
    
    async def get_agent_connections(self, agent_id: str) -> List[AgentConnection]:
        """
        Get all connections for an agent
        Molecular operation: combines connection listing with filtering
        """
        try:
            return await self.agent_service.list_connections(agent_id)
        except Exception as e:
            raise AgentConnectionManagerError(f"Failed to get agent connections: {str(e)}")
    
    async def remove_agent_connection(
        self, 
        agent_id: str, 
        connection_id: str, 
        user_id: str
    ) -> bool:
        """
        Remove agent connection and clean up authentication
        Molecular operation: combines connection removal with auth cleanup
        """
        try:
            # Get connection details before removal for cleanup
            connections = await self.agent_service.list_connections(agent_id)
            connection = next((c for c in connections if c.id == connection_id), None)
            
            # Remove the connection
            removed = await self.agent_service.remove_connection(agent_id, connection_id)
            
            # Clean up authentication if connection was found
            if connection and removed:
                await self.auth_flow_manager.revoke_authentication(user_id, connection.provider_id)
            
            return removed
        except Exception:
            return False
    
    async def refresh_agent_connection(
        self,
        agent_id: str,
        connection_id: str,
        user_id: str
    ) -> AgentConnectionResult:
        """
        Refresh agent connection
        Molecular operation: combines connection refresh with auth refresh
        """
        try:
            # Get current connection details
            connections = await self.agent_service.list_connections(agent_id)
            connection = next((c for c in connections if c.id == connection_id), None)
            
            if not connection:
                return AgentConnectionResult(
                    success=False,
                    error="Connection not found",
                )
            
            # Refresh authentication
            auth_config = AuthFlowConfig(
                provider=connection.provider_id,
                scopes=connection.scopes,
                user_id=user_id,
                agent_id=agent_id,
                is_default=True,
            )
            
            auth_result = await self.auth_flow_manager.refresh_token(auth_config)
            
            if not auth_result.success or auth_result.status != "completed":
                return AgentConnectionResult(
                    success=False,
                    error=auth_result.error or f"Authentication refresh failed with status: {auth_result.status}",
                )
            
            # Remove old connection and create new one
            await self.agent_service.remove_connection(agent_id, connection_id)
            
            connection_request = CreateConnectionRequest(
                oauth_token_id=auth_result.token
            )
            
            new_connection = await self.agent_service.create_connection(agent_id, connection_request)
            
            return AgentConnectionResult(
                success=True,
                connection=new_connection,
            )
            
        except Exception as e:
            return AgentConnectionResult(
                success=False,
                error=f"Failed to refresh agent connection: {str(e)}",
            )
    
    async def get_agent_deployment_status(self, deployment_id: str) -> Optional[AgentDeployment]:
        """
        Get agent deployment status
        Molecular operation: combines deployment retrieval with status checking
        """
        try:
            return await self.agent_service.get_deployment(deployment_id)
        except Exception:
            return None
    
    async def delete_agent(self, agent_id: str, user_id: str) -> bool:
        """
        Delete agent and all its connections
        Molecular operation: combines multiple deletions with cleanup
        """
        try:
            # Get all connections
            connections = await self.agent_service.list_connections(agent_id)
            
            # Remove all connections and clean up auth
            all_connections_removed = True
            for connection in connections:
                removed = await self.remove_agent_connection(agent_id, connection.id, user_id)
                if not removed:
                    all_connections_removed = False
            
            return all_connections_removed
        except Exception:
            return False
    
    async def get_agent_deployments(self) -> List[AgentDeployment]:
        """
        Get all agent deployments
        Molecular operation: combines deployment listing with filtering
        """
        try:
            return await self.agent_service.list_deployments()
        except Exception as e:
            raise AgentConnectionManagerError(f"Failed to get agent deployments: {str(e)}")
    
    async def update_agent_deployment(
        self,
        deployment_id: str,
        config: Dict[str, Any]
    ) -> Optional[AgentDeployment]:
        """
        Update agent deployment
        Molecular operation: combines deployment update with validation
        """
        try:
            return await self.agent_service.patch_deployment(deployment_id, config)
        except Exception:
            return None
    
    async def get_agent_health(self, agent_id: str) -> Dict[str, Any]:
        """
        Get comprehensive agent health status
        Molecular operation: combines multiple health checks
        """
        try:
            connections = await self.get_agent_connections(agent_id)
            
            health_status = {
                "agent_id": agent_id,
                "total_connections": len(connections),
                "healthy_connections": 0,
                "connection_details": [],
                "overall_health": "healthy",
            }
            
            for connection in connections:
                # Check if authentication is still valid
                is_healthy = await self.auth_flow_manager.validate_token(
                    connection.created_by,  # Using created_by as user_id placeholder
                    connection.provider_id
                )
                
                if is_healthy:
                    health_status["healthy_connections"] += 1
                
                health_status["connection_details"].append({
                    "connection_id": connection.id,
                    "provider": connection.provider_id,
                    "healthy": is_healthy,
                    "expires_at": connection.expires_at,
                })
            
            # Determine overall health
            if health_status["healthy_connections"] == 0:
                health_status["overall_health"] = "unhealthy"
            elif health_status["healthy_connections"] < health_status["total_connections"]:
                health_status["overall_health"] = "degraded"
            
            return health_status
            
        except Exception as e:
            return {
                "agent_id": agent_id,
                "overall_health": "error",
                "error": str(e),
            }


# Singleton instance
agent_connection_manager = AgentConnectionManager()
